pallets/free-calls/src/lib.rs

```diff
@@ -0,0 +1,430 @@
+                ));
+            }
+
+            Ok(PostDispatchInfo {
+                actual_weight: Some(actual_weight),
+                pays_fee: Pays::No,
+            })
+        }
+    }
+
+    struct Window<T: Config> {
+        consumer: T::AccountId,
+        config_index: WindowType,
+        config: WindowConfig<T::BlockNumber>,
+        timeline_index: T::BlockNumber,
+        stats: ConsumerStats<T::BlockNumber>,
+        can_have_free_call: bool,
+    }
+
+    impl<T: Config> Window<T> {
+        // TODO: refactor this into more lightweight version??
+        fn build(
+            consumer: T::AccountId,
+            quota: NumberOfCalls,
+            current_block: T::BlockNumber,
+            config_index: WindowType,
+            config: WindowConfig<T::BlockNumber>,
+            window_stats: Option<ConsumerStats<T::BlockNumber>>,
+        ) -> Self {
+            let timeline_index = current_block / config.period;
+
+            let reset_stats = || ConsumerStats::new(timeline_index);
+
+            let mut stats = window_stats.unwrap_or_else(reset_stats);
+
+            if stats.timeline_index < timeline_index {
+                stats = reset_stats();
+            }
+
+            let can_be_called = stats.used_calls < max(1, quota / config.quota_ratio);
+
+            Window {
+                consumer: consumer.clone(),
+                config_index,
+                config,
+                timeline_index,
+                stats,
+                can_have_free_call: can_be_called,
+            }
+        }
+
+        fn increment_window_stats(&mut self) {
+            self.stats.used_calls = self.stats.used_calls.saturating_add(1);
+            <WindowStatsByConsumer<T>>::insert(
+                self.consumer.clone(),
+                self.config_index,
+                self.stats.clone(),
+            );
+        }
+    }
+
+    pub enum ShouldUpdateConsumerStats {
+        YES,
+        NO,
+    }
+
+    impl<T: Config> Pallet<T> {
+        /// Determine if `consumer` can have a free call and optionally update its window usage.
+        ///
+        /// Window usage for the caller `consumer` will only update if there is a quota and all of the
+        /// previous window usages doesn't exceed the defined windows config.
+        pub fn can_make_free_call(consumer: &T::AccountId, should_update_consumer_stats: ShouldUpdateConsumerStats) -> bool {
+            let current_block = <frame_system::Pallet<T>>::block_number();
+
+            let windows_config = T::WindowsConfig::get();
+
+            if windows_config.is_empty() {
+                return false;
+            }
+
+            let locked_info = <LockedInfoByAccount<T>>::get(consumer.clone());
+            let quota = match T::QuotaCalculationStrategy::calculate(current_block, locked_info) {
+                Some(quota) if quota > 0 => quota,
+                _ => return false,
+            };
+
+            let mut windows: Vec<Window<T>> = Vec::new();
```

Comment:
you know the size of this vec will be at most `WindowType` or something, so you should create this vec with the known bound so it doesnt reallocate

Reply:
Ideally you use bounded vecs everywhere for this kind of thing

Reply:
Now all vecs are bounded to the size of the passed in config.

```rust
    /// Retrieves the size of `T::WindowsConfig` to be used for `BoundedVec` declaration.
    pub struct WindowsConfigSize<T: Config>(PhantomData<T>);

    impl<T: Config> Default for WindowsConfigSize<T> {
        fn default() -> Self {
            Self(PhantomData)
        }
    }

    impl<T: Config> Get<u32> for WindowsConfigSize<T> {
        fn get() -> u32 {
            T::WindowsConfig::get().len().try_into().unwrap()
        }
    }
```